ipfs/client.go

```diff
@@ -0,0 +1,167 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package ipfs
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+)
+
+// FileInfo represents the information provided by "/api/v0/files/stat" API of IPFS.
+// Please see details at: https://docs.ipfs.tech/reference/kubo/rpc/#api-v0-files-stat
+type FileInfo struct {
+	Blocks         int    `json:"Blocks"`
+	CumulativeSize uint64 `json:"CumulativeSize": "<uint64>"`
+	Hash           string `json:"Hash": "<string>"`
+	Local          bool   `json:"Local": "<bool>"`
+	Size           uint64 `json:"Size"`
+	SizeLocal      uint64 `json:"SizeLocal"`
+	Type           string `json:"Type"`
+	WithLocality   bool   `json:"WithLocality"`
+}
+
+// StatCID gets and returns information of the file specified by the cid.
+func StatCID(ipfsAPI string, cid string) (info *FileInfo, retErr error) {
+	if ipfsAPI == "" {
+		return nil, fmt.Errorf("specify IPFS API address")
+	}
+	ipfsAPIFilesStat := ipfsAPI + "/api/v0/files/stat"
+	req, err := http.NewRequest("POST", ipfsAPIFilesStat, nil)
+	if err != nil {
+		return nil, err
+	}
+	q := req.URL.Query()
+	q.Add("arg", "/ipfs/"+cid)
+	req.URL.RawQuery = q.Encode()
+	client := http.DefaultClient
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if retErr != nil {
+			io.Copy(io.Discard, resp.Body)
+			resp.Body.Close()
+		}
+	}()
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("failed to stat %v; status code: %v", cid, resp.StatusCode)
+	}
+	var rs FileInfo
+	if err := json.NewDecoder(resp.Body).Decode(&rs); err != nil {
+		return nil, err
+	}
+	return &rs, nil
+}
+
+// Cat get the reader of the data specified by the IPFS path and optionally with
+// the offset and length.
+func Cat(ipfsAPI string, p string, offset *int, length *int) (_ io.ReadCloser, retErr error) {
```

Comment:
Why cat? This doesn't seem to support conCATination

Reply:
Renamed to "Get".